src/main/scala/chiseltest/internal/BackendInterface.scala

```diff
@@ -2,111 +2,93 @@
+    doPeek(signal, new Throwable)
+    val a = simulatorInterface.peek(dataNames(signal)).getOrElse(BigInt(0))
+    logger.debug(s"${resolveName(signal)} -> $a")
+    a > 0
+  }
 
   /** Writes a value to a writable wire.
     * Throws an exception if write is not writable.
     */
-  def pokeBits(signal: Data, value: BigInt): Unit
+  def pokeBits(signal: Data, value: BigInt): Unit = {
+    doPoke(signal, value, new Throwable)
+    val dataName = dataNames(signal)
+    simulatorInterface.peek(dataName) match {
+      case Some(peekValue) =>
+        if (peekValue != value) {
+          idleCycles.clear()
+        }
+        simulatorInterface.poke(dataNames(signal), value)
+        logger.debug(s"${resolveName(signal)} <- $value")
+      case None =>
+        logger.debug(s"${resolveName(signal)} is eliminated by firrtl, ignore it. ")
+    }
+  }
 
   /** Returns the current value on a wire.
     * If stale is true, returns the current combinational value (after previous pokes have taken effect).
     * If stale is false, returns the value at the beginning of the current cycle.
     */
-  def peekBits(signal: Data, stale: Boolean): BigInt
-
-  def expectBits(
-    signal:  Data,
-    value:   BigInt,
-    message: Option[String],
-    decode:  Option[BigInt => String],
-    stale:   Boolean
-  ): Unit
-
-  /**
-    * Sets the timeout of the clock: the number of cycles the clock can advance without
-    * some non-nop poke operation.
-    * Setting cycles=0 disables the timeout.
-    * Setting cycles=1 means every cycle must have some non-nop poke operation.
-    * Resets the idle counter associated with the specified clock.
-    */
-  def setTimeout(signal: Clock, cycles: Int): Unit
-
-  /** Advances the target clock by one cycle.
-    */
-  def step(signal: Clock, cycles: Int): Unit
-
-  def doFork(runnable: () => Unit, name: Option[String], region: Option[Region]): AbstractTesterThread
-
-  def doJoin(threads: Seq[AbstractTesterThread], stepAfter: Option[Clock]): Unit
-
-  def doTimescope(contents: () => Unit): Unit
-
-  // Circuit introspection functionality
-  //
-  /** Returns set of clocks associated with sources of the signal
-    */
-  def getSourceClocks(signal: Data): Set[Clock]
-
-  /** Returns set of clocks associated with sinks of the signal
-    */
-  def getSinkClocks(signal: Data): Set[Clock]
-
-  // Test Instance State
-  //
-
-  /** Returns the stack trace elements of parent threads. If currently in the root thread, returns
-    * empty.
-    * TODO: refactor this, figure out how to do this in a structurally cleaner way
-    */
-  def getParentTraceElements: Seq[StackTraceElement] = Seq()
-
-  protected val testMap = mutable.HashMap[Any, Any]()
-
-  /** Sets the value associated with a key in a per-test map.
-    */
-  def setVar(key: Any, value: Any): Unit = {
-    testMap.put(key, value)
+  def peekBits(signal: Data, stale: Boolean): BigInt = {
+    require(!stale, "Stale peek not yet implemented")
```

Comment:
Feel free to remove the stale stuff. I don't think it's going to happen (or that it even was a good idea).

Reply:
all stale removed.